ehrapy/tools/cohort_tracking/_cohort_tracker.py

```diff
@@ -0,0 +1,332 @@
+        self.track = self._get_column_structure(df)
+
+        self._track_backup = copy.deepcopy(self.track)
+
+    def __call__(
+        self, adata: AnnData, label: str = None, operations_done: str = None, *args: Any, **tableone_kwargs: Any
+    ) -> Any:
+        if isinstance(adata, AnnData):
+            df = adata.obs
+        elif isinstance(adata, pd.DataFrame):
+            df = adata
+        else:
+            raise ValueError("adata must be an AnnData or a DataFrame.")
+
+        _check_columns_exist(df, self.columns)
+
+        # track a small text with each tracking step, for the flowchart
+        track_text = label if label is not None else f"Cohort {self.tracked_steps}"
+        track_text += "\n (n=" + str(adata.n_obs) + ")"
+        self._tracked_text.append(track_text)
+
+        # track a small text with the operations done
+        self._tracked_operations.append(operations_done)
+
+        self._tracked_steps += 1
+
+        t1 = TableOne(df, columns=self.columns, categorical=self.categorical, **tableone_kwargs)
+        # track new stuff
+        self._get_column_dicts(t1)
+
+    def _get_column_structure(self, df):
+        column_structure = {}
+        for column in self.columns:
+            if column in self.categorical:
+                # if e.g. a column containing integers is deemed categorical, coerce it to categorical
+                df[column] = df[column].astype("category")
+                column_structure[column] = {category: [] for category in df[column].cat.categories}
+            else:
+                column_structure[column] = []
+
+        return column_structure
+
+    def _get_column_dicts(self, table_one):
+        for col, value in self.track.items():
+            if isinstance(value, dict):
+                self._get_cat_dicts(table_one, col)
+            else:
+                self._get_num_dicts(table_one, col)
+
+    def _get_cat_dicts(self, table_one, col):
+        for cat in self.track[col].keys():
+            # if tableone does not have the category of this column anymore, set the percentage to 0
+            # for categorized columns (e.g. gender 1.0/0.0), str(cat) helps to avoid considering the category as a float
+            if (col, str(cat)) in table_one.cat_table["Overall"].index:
+                pct = float(table_one.cat_table["Overall"].loc[(col, str(cat))].split("(")[1].split(")")[0])
+            else:
+                pct = 0
+            self.track[col][cat].append(pct)
+
+    def _get_num_dicts(self, table_one, col):
+        summary = table_one.cont_table["Overall"].loc[(col, "")]
+        self.track[col].append(summary)
+
+    def reset(self):
+        self.track = self._track_backup
+        self._tracked_steps = 0
+        self._tracked_text = []
+        self._tracked_operations = []
+
+    @property
+    def tracked_steps(self):
+        return self._tracked_steps
+
+    def plot_cohort_change(
```

Comment:
Always good if one allows the passing of ax

Reply:
The heterogeneity in how scanpy and ehrapy take plotting arguments and use them raises the question to what flavor we settle here. (e.g. in `sc.pp.embedding`, an `ax` is taken, but `return_fig=True` then yields an error b.c. in that case a fig is never created, and so on)
Probably needs harmonization in future, so going for simple & useful for now?
Suggest using a `show` then instead of `return_figure`, as ep.pl.kmf and ep.pl.ols do: if `show=True`, show figure, if `show=False`, return `ax`

Reply:
There's an issue for this on the pertpy repository and @namsaraeva is looking into that. We'll also need this fixed for ehrapy. Gregor made great suggestions on zulip that I think we should eventually follow. This is probably a different and slightly bigger issues so you don't need to worry about that here.